### PR TITLE
[spi_device] TPM Read to wait full payload before send

### PR DESCRIPTION
**Pull request by eunchan**
This commit is a follow-up of https://github.com/lowRISC/opentitan/pull/14544

TPM, now, waits for the read FIFO having enough data.

### CODE REVIEW

File: hw/ip/spi_device/rtl/spi_tpm.sv
Comment (eunchan) [inline hw/ip/spi_device/rtl/spi_tpm.sv:769]: Hum,,, I will try to find better coding here. I don't like this. :(

Comment (eunchan) [reply hw/ip/spi_device/rtl/spi_tpm.sv]: Updated.

File: hw/ip/spi_device/rtl/spi_tpm.sv
Comment (a-will) [inline hw/ip/spi_device/rtl/spi_tpm.sv:789]: For what it's worth, I think this section is not needed (except for needing to release from the wait state where it's impossible for rdepth to cover the full xfer_size). I'd argue that we cannot claim to support an xfer_size that is larger than the FIFO. There are too many ways this could go wrong at the system level--In general, "streaming" over synchronous interfaces with software in the loop is not advisable.

Comment (a-will) [reply hw/ip/spi_device/rtl/spi_tpm.sv:789]: And by that, I mean that this could collapse into just...

```systemverilog
      enough_payload_in_rdfifo = ({isck_rdfifo_rdepth, RdFifoBytesW'(0)} > {1'b 0, xfer_size}) |
                                 (RdFifoSize <= {1'b0, xfer_size});
```

The comments suggesting streaming through software invite a problematic use case.

Comment (eunchan) [reply hw/ip/spi_device/rtl/spi_tpm.sv:789]: Acked. I will remove that part from the design and spec also. @tjaychen please comment if limiting the supported transfer size to the actual FIFO size is good or not :)

**Comment by eunchan** [comment]
Rebased to #14579

**Comment by eunchan** [comment]
Looks like Private CI has been completed successfully.

image

But GitHub couldn't catch it. CC: @GregAC

**Comment by GregAC** [comment]
> Looks like Private CI has been completed successfully.
> 
> image
> 
> But GitHub couldn't catch it. CC: @GregAC

Hmm odd, I've tried rerunning the whole private CI job, that might solve it. Otherwise I'd suggest rebasing on master (just so the commit hashes change) and force pushing the new version to trigger a new CI run.

**Comment by eunchan** [comment]
> Hmm odd, I've tried rerunning the whole private CI job, that might solve it. Otherwise I'd suggest rebasing on master (just so the commit hashes change) and force pushing the new version to trigger a new CI run.
>
OK. But i don't think it is worth to re-run CI . As it shows the CI (private) completed without any errors, I suggest merging this as it is (after getting approvals)

**Review by a-will — APPROVED** [review]
This LGTM, but note that the second condition was dropped, so if the read FIFO is less than 64 B in some future iteration, it looks like the TPM module can freeze in the wait state.

**Comment by eunchan** [comment]
> This LGTM, but note that the second condition was dropped, so if the read FIFO is less than 64 B in some future iteration, it looks like the TPM module can freeze in the wait state.
>
Ah, you are right. It should anyhow flush the content. I don't think HW should report an error as SW already received the transfer size via CMDFIFO. Anyhow, let me revise.

File: hw/ip/spi_device/data/spi_device.hjson
Comment (a-will) [inline hw/ip/spi_device/data/spi_device.hjson:1073]: Oh, I just noticed: Was this blurb supposed to go in the "max_rd_size" section? It looks like these bits are for "max_wr_size"

Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:1073]: I do not know where I addressed this. But should be in somewhere in other PRs. Hum... I could not find it .. Maybe it would be disappeared while I was rebasing.. let me quick-fix it

Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:1073]: shoot. I didn't upload the commit :( Let me amend here.

Comment (a-will) [reply hw/ip/spi_device/data/spi_device.hjson:1073]: Thanks, @eunchan!

**Pull request merged**
